ambry-replication/src/main/java/com/github/ambry/replication/ReplicationManager.java

```diff
@@ -451,4 +487,197 @@ public void onPartitionBecomeDroppedFromOffline(String partitionName) {
+     */
+    public Map<String, Set<ReplicaId>> getPeerLeaderReplicasByPartition() {
+      return partitionLeaderInfo.getPeerLeaderReplicasByPartition();
+    }
+
+    /**
+     * Checks if a remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+     * @param partitionName name of local leader partition
+     * @param replicaId remote replica to be checked
+     * @return true if remote replica is a leader for a partition (Pre-requisite: the partition itself should be a leader locally).
+     */
+    public boolean isPeerReplicaLeaderForPartition(String partitionName, ReplicaId replicaId) {
+      return partitionLeaderInfo.isPeerReplicaLeaderForPartition(partitionName, replicaId);
+    }
+
+    /**
+     * Maintains the list of leader partitions on local node and their corresponding peer leaders in remote data centers
+     */
+    class PartitionLeaderInfo {
+
+      private final Map<String, Set<ReplicaId>> peerLeaderReplicasByPartition = new ConcurrentHashMap<>();
+      private final ReadWriteLock rwLockForLeaderReplicaUpdates = new ReentrantReadWriteLock();
+
+      public PartitionLeaderInfo() {
+
+        // We can't initialize the peerLeaderReplicasByPartition map on startup because we don't know the leader partitions
+        // on local server until it has finished participating with Helix. It will be updated after server participates
+        // with Helix and receives LEADER transition (via onPartitionBecomeLeaderFromStandby()).
+      }
+
+      /**
+       * Get a map of partitions to their sets of peer leader replicas (this method is only by ReplicationTest for now)
+       * @return an unmodifiable map of peer leader replicas stored by partition
+       */
+      public Map<String, Set<ReplicaId>> getPeerLeaderReplicasByPartition() {
+        return Collections.unmodifiableMap(peerLeaderReplicasByPartition);
+      }
+
+      /**
+       * Add a leader partition and its set of peer leader replicas. This method is thread safe.
+       * @param partitionName name of the partition to be added
+       */
+      public void addPartition(String partitionName) {
+
+        // 1. get local replica from store manager
+        ReplicaId localReplica = storeManager.getReplica(partitionName);
+
+        // Read-write lock avoids contention from threads removing old leader partitions (removePartition()) and
+        // threads updating existing leader partitions (refreshPeerLeadersForAllPartitions())
+        rwLockForLeaderReplicaUpdates.writeLock().lock();
+        try {
+          // 2. Get the peer leader replicas from all data centers for this partition
+          List<? extends ReplicaId> leaderReplicas =
+              localReplica.getPartitionId().getReplicaIdsByState(ReplicaState.LEADER, null);
+
+          // 3. Collect and log the list of peer leader replicas associated with this partition
+          Set<ReplicaId> peerLeaderReplicas = new HashSet<>();
```

Comment:
It looks like `refreshPeerLeadersForAllPartitions` has a slightly different way of building the set of peerLeaderReplicas. Could both of these methods use the same approach (whichever one you prefer)?

Reply:
Yeah, true. Made changes to keep the approach same in both methods.